File: typescript/examples/agents/05-hooks-timing.ts
// Hooks for performance measurement — track LLM latency and tool execution time
// This pattern is useful for building dashboards or sending metrics to your observability stack.
import { z } from "zod";
import { Agent, tool } from "../../src/index.js";
import type { Hooks } from "../../src/index.js";

const getWeather = tool({
  name: "get_weather",
  description: "Get the current weather for a city",
  parameters: z.object({
    city: z.string().describe("City name"),
  }),
  execute: async ({ city }) => {
    // Simulate variable latency
    await new Promise((r) => setTimeout(r, 50 + Math.random() * 100));
    return {
      city,
      temperature: Math.round(15 + Math.random() * 15),
      condition: ["sunny", "cloudy", "rainy"][Math.floor(Math.random() * 3)],
    };
  },
});

// Collect timing metrics in a structured way
interface Metrics {
  llmCalls: Array<{ iteration: number; durationMs: number }>;
  toolCalls: Array<{ name: string; durationMs: number }>;
  totalDurationMs: number;
}

function createTimingHooks(): { hooks: Hooks; metrics: Metrics } {
  const metrics: Metrics = { llmCalls: [], toolCalls: [], totalDurationMs: 0 };
  const timers = new Map<string, number>();

  const hooks: Hooks = {
    onAgentStart: () => {
      timers.set("agent", Date.now());
    },
    onAgentEnd: () => {
      metrics.totalDurationMs = Date.now() - (timers.get("agent") ?? Date.now());
    },
    onLLMCall: ({ iteration }) => {
      timers.set(`llm:${iteration}`, Date.now());
    },
    onLLMResponse: ({ iteration }) => {
      const start = timers.get(`llm:${iteration}`);
      if (start) {
        metrics.llmCalls.push({ iteration, durationMs: Date.now() - start });
      }
    },
    onToolEnd: ({ name, durationMs }) => {
      metrics.toolCalls.push({ name, durationMs });
    },
  };

  return { hooks, metrics };
}

const { hooks, metrics } = createTimingHooks();

const agent = new Agent({
  name: "weather-agent",
  instructions: "You help users check the weather. Use the get_weather tool. Be concise.",
  tools: [getWeather],
  hooks,
});

const result = await agent.run("What's the weather in Berlin?");

// Print structured metrics
console.log("=== Performance Metrics ===\n");
console.log(`Total duration: ${metrics.totalDurationMs}ms`);
console.log(`Iterations: ${result.meta.iterations}`);
console.log(`\nLLM calls:`);
for (const call of metrics.llmCalls) {
  console.log(`  Iteration ${call.iteration}: ${call.durationMs}ms`);
}
console.log(`\nTool calls:`);
for (const call of metrics.toolCalls) {
  console.log(`  ${call.name}: ${call.durationMs}ms`);
}
console.log(`\nToken usage:`);
console.log(`  Input:  ${result.meta.usage.inputTokens}`);
console.log(`  Output: ${result.meta.usage.outputTokens}`);
console.log(`  Total:  ${result.meta.usage.totalTokens}`);
console.log(`\nAnswer: ${result.output}`);

File: python/examples/agents/05_hooks_timing.py
# Hooks for timing — measure LLM latency and tool execution time

import asyncio
import time

from opperai.agent import Agent, Hooks, tool


@tool
def slow_lookup(query: str) -> str:
    """Simulate a slow database lookup."""
    time.sleep(0.1)  # Simulate latency
    return f"Results for: {query}"


# Collect metrics using closure state
metrics: dict = {
    "llm_calls": 0,
    "llm_total_ms": 0,
    "tool_total_ms": 0,
    "tools_called": 0,
}
_llm_start: dict = {}


def on_llm_call(ctx: dict) -> None:
    _llm_start[ctx["iteration"]] = time.monotonic()


def on_llm_response(ctx: dict) -> None:
    start = _llm_start.pop(ctx["iteration"], None)
    if start:
        elapsed_ms = (time.monotonic() - start) * 1000
        metrics["llm_calls"] += 1
        metrics["llm_total_ms"] += elapsed_ms
        print(f"  LLM call #{metrics['llm_calls']}: {elapsed_ms:.0f}ms")


def on_tool_end(ctx: dict) -> None:
    metrics["tools_called"] += 1
    metrics["tool_total_ms"] += ctx["duration_ms"]
    print(f"  Tool {ctx['name']}: {ctx['duration_ms']:.0f}ms")


async def main() -> None:
    hooks = Hooks(
        on_llm_call=on_llm_call,
        on_llm_response=on_llm_response,
        on_tool_end=on_tool_end,
    )

    agent = Agent(
        name="timed-agent",
        instructions="You help with lookups. Be concise.",
        tools=[slow_lookup],
        hooks=hooks,
    )

    result = await agent.run("Look up 'machine learning' and 'deep learning'")

    print("\n--- Performance Summary ---")
    print(
        f"LLM calls: {metrics['llm_calls']} ({metrics['llm_total_ms']:.0f}ms total)"
    )
    print(
        f"Tool calls: {metrics['tools_called']} ({metrics['tool_total_ms']:.0f}ms total)"
    )
    print(f"Tokens: {result.meta.usage.total_tokens}")


if __name__ == "__main__":
    asyncio.run(main())
